app/controllers/declaration/DeclarationAdditionalActorsController.scala

```diff
@@ -82,22 +78,19 @@ class DeclarationAdditionalActorsController @Inject()(
     implicit request: JourneyRequest[AnyContent],
     hc: HeaderCarrier
   ): Future[Result] =
-    (userInput, cachedData.actors) match {
-      case (_, actors) if actors.length >= maxNumberOfItems =>
-        handleErrorPage(mode, Seq(("", exceedMaximumNumberError)), userInput, cachedData.actors)
-
-      case (actor, actors) if actors.contains(actor) =>
-        handleErrorPage(mode, Seq(("", duplicateActorError)), userInput, cachedData.actors)
-
-      case (actor, actors) =>
-        if (actor.isDefined) {
-          val updatedCache = DeclarationAdditionalActorsData(actors :+ actor)
-          updateCache(updatedCache)
-            .map(_ => navigator.continueTo(mode, routes.DeclarationAdditionalActorsController.displayPage))
-        } else
-          handleErrorPage(mode, Seq(("eori", "supplementary.additionalActors.eori.isNotDefined")), userInput, cachedData.actors)
+    actorsValidator(mode, userInput, cachedData.actors) {
+      val updatedCache = DeclarationAdditionalActorsData(cachedData.actors).addActor(userInput)
+      updateCache(updatedCache)
+        .map(_ => navigator.continueTo(mode, routes.DeclarationAdditionalActorsController.displayPage))
     }
 
+  private def actorsValidator(mode: Mode, actor: DeclarationAdditionalActors, actors: Seq[DeclarationAdditionalActors])(
+    success: Future[Result]
+  )(implicit request: JourneyRequest[AnyContent], hc: HeaderCarrier): Future[Result] =
+    DeclarationAdditionalActors
+      .actorsValidator(actor, actors)
+      .fold[Future[Result]](success)(errors => handleErrorPage(mode, errors, actor, actors))
```

Comment:
👍 Nice!
I could only suggest a different naming for parameters: actors -> existingActors (?), success -> onSuccess